csharp/src/Google.Protobuf/Reflection/CustomOptions.cs

```diff
@@ -205,98 +232,116 @@ internal CustomOptions(IDictionary<int, IExtensionValue> values)
+                {
+                    if (extensionValue is ExtensionValue<T>)
+                    {
+                        ExtensionValue<T> single = extensionValue as ExtensionValue<T>;
+                        ByteString bytes = single.GetValue().ToByteString();
+                        value = new T();
+                        value.MergeFrom(bytes);
+                        return true;
+                    }
+                    else if (extensionValue is RepeatedExtensionValue<T>)
+                    {
+                        RepeatedExtensionValue<T> repeated = extensionValue as RepeatedExtensionValue<T>;
+                        value = repeated.GetValue()
+                            .Select(v => v.ToByteString())
+                            .Aggregate(new T(), (t, b) =>
+                            {
+                                t.MergeFrom(b);
+                                return t;
+                            });
+                        return true;
+                    }
+                }
             }
 
-            IExtensionValue extensionValue;
-            if (values.TryGetValue(field, out extensionValue))
+            UnknownField f;
+            if (UnknownFieldSet.TryGetField(unknownFields, field, out f))
             {
-                if (extensionValue is ExtensionValue<T>)
+                value = new T();
+                foreach (var bytes in f.LengthDelimitedList ?? Enumerable.Empty<ByteString>())
                 {
-                    ExtensionValue<T> single = extensionValue as ExtensionValue<T>;
-                    ByteString bytes = single.GetValue().ToByteString();
-                    value = new T();
                     value.MergeFrom(bytes);
-                    return true;
-                }
-                else if (extensionValue is RepeatedExtensionValue<T>)
-                {
-                    RepeatedExtensionValue<T> repeated = extensionValue as RepeatedExtensionValue<T>;
-                    value = repeated.GetValue()
-                        .Select(v => v.ToByteString())
-                        .Aggregate(new T(), (t, b) =>
-                        {
-                            t.MergeFrom(b);
-                            return t;
-                        });
-                    return true;
                 }
+                return true;
             }
 
             value = null;
             return false;
         }
 
-        private bool TryGetPrimitiveValue<T>(int field, out T value)
-        {
-            if (values == null)
-            {
-                value = default(T);
-                return false;
-            }
+        private delegate bool TryGetLastValue<T>(UnknownField field, out T value);
 
-            IExtensionValue extensionValue;
-            if (values.TryGetValue(field, out extensionValue))
+        private bool TryGetPrimitiveValue<T, U>(int field, TryGetLastValue<U> getValue, Func<U, T> conv, out T value)
```

Comment:
it feels odd that you're passing a further unspecified "getValue" function, but you're only using it only sometimes, when reading from the unknownFieldSet - feels dirty.
Same for the "conv" argument.

Reply:
Here's a idea how to refactor for improved clarity

```
public bool TryGetInt32(int field, out int value)
{
     if (TryGetPrimitiveValueFromExtensions())
     {
         return true;
     }
     // the method's summary would have an explanation why it's also important to check the unknown fields
    // when calling this method, the "getValue" and "conv" arguments make more sense.
     if (TryGetPrimitiveValueFromUnknownFields(filed, getValueFunc, conversionFunc))
     {
        // alternatively, you can perform the conversion here...
       return true;
     }
    return false;
}
```

Reply:
Adding that to *every* method would add a lot of bloat that could probably be handled with a good doc comment and comments throughout the function. For example, something like...
```cs
/// <summary>
/// Tries to get the primitive value from the extension dictionary, otherwise falls-back to the unknown field set.
/// </summary>
/// <typeparam name="T">The returned type</typeparam>
/// <typeparam name="U">The unknown field representation</typeparam>
/// <param name="field">The field number to try and get a value for</param>
/// <param name="getValue">A function that tries to get the last value of a field of a specific type in an unknown field.</param>
/// <param name="conv">A conversion function that converts from an unknown field value to the output value</param>
/// <param name="value">The resulting value</param>
/// <returns><c>true</c> if a field value was found, <c>false</c> otherwise</returns>
private bool TryGetPrimitiveValue<T, U>(int field, TryGetLastValue<U> getValue, Func<U, T> conv, out T value)
```

Reply:
Though now that I look at it it seems you may still the 2 calls relatively small with an `||`
```
public bool TryGetInt32(int field, out int value)
    => TryGetExtensionValue<int>(field, out value)
    || TryGetLastUnknownValue<int, ulong>(field, UnknownField.TryGetLastVarint, (v) => (int)v, out value);
```
I don't know if it really looks any better. I guess I could see how it makes it more explicit up front...